sdk/src/transaction_context.rs

```diff
@@ -682,19 +810,101 @@ impl<'a> BorrowedAccount<'a> {
+        {
+            return Ok(());
+        }
+        let not_owner = !self.is_owned_by_current_program();
+        if new_length != self.get_data().len() {
+            // Accounts data meter
+            let total_resize_delta = self
+                .transaction_context
+                .total_resize_delta
+                .borrow()
+                .saturating_add((new_length as i64).saturating_sub(self.get_data().len() as i64));
+            if total_resize_delta > self.transaction_context.total_resize_limit as i64 {
+                return Err(InstructionError::MaxAccountsDataSizeExceeded);
+            }
+            // Only the owner can change the length of the data
+            if not_owner {
+                return Err(InstructionError::AccountDataSizeChanged);
+            }
+        }
+        // The new length can not exceed the maximum permitted length
+        if new_length > MAX_PERMITTED_DATA_LENGTH as usize {
+            return Err(InstructionError::InvalidRealloc);
+        }
+        // Only non-executable accounts data can be changed
+        if self.is_executable() {
+            return Err(InstructionError::ExecutableDataModified);
+        }
+        // and only if the account is writable
+        if !self.is_writable() {
+            return Err(InstructionError::ReadonlyDataModified);
+        }
+        // and only if we are the owner
+        if not_owner {
+            return Err(InstructionError::ExternalAccountDataModified);
+        }
+        Ok(())
+    }
+
+    fn touch(&self) -> Result<(), InstructionError> {
+        *self
+            .transaction_context
+            .account_touched_flags
+            .borrow_mut()
+            .get_mut(self.index_in_transaction)
+            .ok_or(InstructionError::NotEnoughAccountKeys)? = true;
+        Ok(())
+    }
 }
 
 /// Everything that needs to be recorded from a TransactionContext after execution
 pub struct ExecutionRecord {
     pub accounts: Vec<TransactionAccount>,
     pub instruction_trace: InstructionTrace,
     pub return_data: TransactionReturnData,
+    pub changed_account_count: u64,
+    pub total_size_of_all_accounts: u64,
+    pub total_size_of_touched_accounts: u64,
     pub total_resize_delta: i64,
 }
 
 /// Used by the bank in the runtime to write back the processed accounts and recorded instructions
 impl From<TransactionContext> for ExecutionRecord {
     fn from(context: TransactionContext) -> Self {
+        let mut changed_account_count = 0u64;
+        let mut total_size_of_all_accounts = 0u64;
+        let mut total_size_of_touched_accounts = 0u64;
+        for (index_in_transaction, flag) in
+            context.account_touched_flags.borrow().iter().enumerate()
+        {
+            let account_data_size = context
+                .get_account_at_index(index_in_transaction)
+                .map(|account| account.borrow().data().len())
+                .unwrap_or(0) as u64;
```

Comment:
nit: i think the `_or(0)` part should never hit in production, unless something very smelly is ongogigng, right? in that case, I'd recommend the hard `.expect("no out-of-range index_in_transaction")` or similar, because we're already introducing a new panic source in `.account_touched_flags.{borrow,borrow_mut}()`